Cython/Utility/CConvert.pyx

```diff
@@ -130,3 +130,69 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
         Py_INCREF(value)
         PyTuple_SET_ITEM(t, i, value)
     return t
+
+#################### c_array_slice.to_py ####################
```

Comment:
Could you largely re-use the `carray.to_py` section? Your functions would just have to work out a start and end index (taking into account wraparound) and then call the existing convension functions?

I'm also not convinced we really need a set/frozenset version - these are probably used rarely enough that they could just go through a tuple.

Reply:
In https://github.com/cython/cython/pull/3151#discussion_r425810643 to support list, tuple, set, frozenset

Reply:
I missed that comment - sorry!

Note that you can add use [`PySet_Add`](https://docs.python.org/3/c-api/set.html#c.PySet_Add) on frozenset's too - you don't have to go through a tuple.

Reply:
I used it before, but it fails in PyPy for frozenset, so I have used
__Pyx_PyFrozenSet_New
by line
```
{{if to_py_obj_type == 'frozenset'}}
l = frozenset(l)
{{endif}}
```
by converting from tuple

Reply:
I've opened a [PyPy bug for this](https://foss.heptapod.net/pypy/pypy/-/issues/3251). You should probably keep your workaround (with appropriate reference counting fixes) rather than wait for a fix from PyPy.

Reply:
You could also just declare `__Pyx_PyFrozenSet_New(PyObject* it)` and call it as `__Pyx_PyFrozenSet_New(NULL)`. It seems to support that. (You would then have to call `use_utility_code()` for the `pyfrozenset_new` utility in the `frozenset` case.)